File: eve_mcp/tools/account.py
"""Session state and the at-a-glance character overview."""
from __future__ import annotations

import asyncio
from typing import Any

from mcp.server.mcpserver import MCPServer

from ..config import CORP_READ_SCOPES, WRITE_CAPABILITIES
from ..context import AppContext
from ..esi import EsiError
from ..schema import CharacterArg
from ._common import handled, isk

_ALL_WRITE_SCOPES = {s for cap in WRITE_CAPABILITIES.values() for s in cap.scopes}


def register(mcp: MCPServer, ctx: AppContext) -> None:
    @mcp.tool()
    @handled
    async def eve_server_status() -> dict[str, Any]:
        """Tranquility server status: player count, build version, uptime, VIP mode.

        Also the cheapest way to confirm this server can reach ESI at all. EVE
        has a daily downtime around 11:00 UTC; a low player count right after it
        is normal, not a bug.

        Returns: server_version, players, vip, start_time, data_age.
        """
        result = await ctx.esi.get("/status")
        return {**result.data, "data_age": result.stale_note}

    @mcp.tool()
    @handled
    async def eve_auth_status() -> dict[str, Any]:
        """Who is authorized here, and which in-game changes this server permits.

        Call this before anything else when you do not know the setup, and
        always before promising the user an in-game change. It answers three
        questions at once: which characters you can query, which mutating
        capabilities the operator enabled, and how much of the hourly write
        budget is left.

        Returns: characters[], default_character, write_mode, enabled_capabilities,
        disabled_capabilities, capability_reference, corporation_scopes_requested,
        budgets, audit_log.
        """
        tokens = ctx.sso.store.all()
        policy = ctx.guard.status()
        policy["outward_facing_capabilities"] = sorted(
            name for name, cap in WRITE_CAPABILITIES.items() if cap.outward_facing
        )
        policy["write_mode_meaning"] = {
            "off": "no writes at all; write scopes are never even requested at login",
            "confirm": "each write returns a preview plus a one-time token; a second call executes it",
            "on": "writes execute immediately (still budgeted and audited)",
        }[ctx.settings.write_mode]

        if not tokens:
            return {
                "characters": [],
                "next_step": (
                    "Nobody is authorized. Call eve_auth_login_url and give the user "
                    "the link to open in a browser."
                ),
                "corporation_scopes_requested": ctx.settings.corp_scopes,
                **policy,
            }
        return {
            "characters": [
                {
                    "name": t.character_name,
                    "character_id": t.character_id,
                    "scope_count": len(t.scopes),
                    # Derive from the capability model, not a substring match: the
                    # old filter silently missed esi-mail.organize_mail.v1 and
                    # esi-calendar.respond_calendar_events.v1.
                    "write_scopes": sorted(set(t.scopes) & _ALL_WRITE_SCOPES),
                    "corporation_scopes": sorted(set(t.scopes) & set(CORP_READ_SCOPES)),
                }
                for t in tokens
            ],
            "default_character": tokens[0].character_name if len(tokens) == 1 else None,
            "corporation_scopes_requested": ctx.settings.corp_scopes,
            **policy,
        }

    @mcp.tool()
    @handled
    async def eve_auth_login_url() -> dict[str, Any]:
        """Generate an EVE SSO link the user must open to authorize a character.

        You cannot complete this yourself — hand the URL to the user. They log
        in with their EVE account, approve the scope list, and the server stores
        the resulting token. One-time per character; several characters can be
        authorized by repeating it.

        Returns: login_url, scope_count, write_capabilities_requested,
        corporation_scopes_requested, instructions.
        """
        url, state = ctx.sso.build_login()
        scopes = ctx.settings.requested_scopes()
        return {
            "login_url": url,
            "state": state,
            "scope_count": len(scopes),
            "write_capabilities_requested": sorted(ctx.settings.write_allow)
            if ctx.settings.write_mode != "off"
            else [],
            "corporation_scopes_requested": ctx.settings.corp_scopes,
            "instructions": (
                "Open login_url in a browser, pick the character, approve. "
                "The link is valid for 15 minutes and works once."
            ),
        }

    @mcp.tool()
    @handled
    async def eve_auth_logout(character: CharacterArg) -> dict[str, Any]:
        """Revoke this server's access to one character and delete its stored token.

        Irreversible in the sense that re-authorizing needs another browser
        login, but it destroys nothing in-game.

        Returns: removed, character_id.
        """
        token = ctx.resolve_character(character)
        await ctx.sso.revoke(token.character_id)
        return {"removed": token.character_name, "character_id": token.character_id}

    @mcp.tool()
    @handled
    async def eve_character_overview(character: CharacterArg = "") -> dict[str, Any]:
        """Everything you would glance at on logging in: corp, ISK, location, ship, training.

        The best first call for almost any question about how the character is
        doing — it fuses seven ESI endpoints into roughly 200 tokens and tells
        you what to drill into next. It already includes the wallet balance and
        what is training, so there is no need to ask for those separately.

        Partial results are normal: if one underlying endpoint fails, the rest
        still come back rather than the whole call erroring.

        Returns: name, corporation, alliance, security_status, wallet_isk,
        online, solar_system, docked_at, ship_type, training_now, queue_ends,
        remaps_available. `location_age` is separate because location is cached
        for only 5 seconds while other fields are cached far longer.
        """
        token = ctx.resolve_character(character)
        cid = token.character_id

        async def maybe(coro):
            try:
                return await coro
            except (EsiError, Exception):  # noqa: BLE001 - partial overview beats none
                return None

        public, wallet, location, ship, online, queue, attributes = await asyncio.gather(
            maybe(ctx.esi.get(f"/characters/{cid}")),
            maybe(ctx.esi.get(f"/characters/{cid}/wallet", character_id=cid)),
            maybe(ctx.esi.get(f"/characters/{cid}/location", character_id=cid)),
            maybe(ctx.esi.get(f"/characters/{cid}/ship", character_id=cid)),
            maybe(ctx.esi.get(f"/characters/{cid}/online", character_id=cid)),
            maybe(ctx.esi.get(f"/characters/{cid}/skillqueue", character_id=cid)),
            maybe(ctx.esi.get(f"/characters/{cid}/attributes", character_id=cid)),
        )

        out: dict[str, Any] = {"character_id": cid, "name": token.character_name}

        if public is not None:
            info = public.data
            ids = [i for i in (info.get("corporation_id"), info.get("alliance_id")) if i]
            names = await ctx.resolver.names(ids)
            out["corporation"] = names.get(info.get("corporation_id"))
            if info.get("alliance_id"):
                out["alliance"] = names.get(info["alliance_id"])
            out["security_status"] = round(info.get("security_status", 0.0), 2)
            out["birthday"] = info.get("birthday")

        if wallet is not None:
            out["wallet_isk"] = wallet.data
            out["wallet"] = isk(wallet.data)

        if online is not None:
            out["online"] = online.data.get("online")
            out["last_login"] = online.data.get("last_login")

        if location is not None:
            loc = location.data
            place_ids = [
                i
                for i in (loc.get("solar_system_id"), loc.get("station_id"), loc.get("structure_id"))
                if i
            ]
            names = await ctx.resolver.names(place_ids, character_id=cid)
            out["solar_system"] = names.get(loc.get("solar_system_id"))
            docked_at = loc.get("station_id") or loc.get("structure_id")
            out["docked_at"] = names.get(docked_at) if docked_at else "in space"
            out["location_age"] = location.stale_note

        if ship is not None:
            out["ship_type"] = await ctx.resolver.name(ship.data.get("ship_type_id", 0))
            ship_name = ship.data.get("ship_name")
            if ship_name and ship_name != out["ship_type"]:
                out["ship_name"] = ship_name

        if queue is not None:
            entries = [e for e in queue.data or [] if e.get("finish_date")]
            if entries:
                first = entries[0]
                skill_name = await ctx.resolver.name(first["skill_id"])
                out["training_now"] = f"{skill_name} {_roman(first.get('finished_level'))}"
                out["training_finishes"] = first.get("finish_date")
                out["queue_length"] = len(entries)
                out["queue_ends"] = entries[-1].get("finish_date")
            else:
                out["training_now"] = None
                out["warning"] = "Skill queue is empty — training time is being wasted."

        if attributes is not None:
            out["remaps_available"] = attributes.data.get("bonus_remaps")

        return {k: v for k, v in out.items() if v not in (None, "", [], {})}


def _roman(level: int | None) -> str:
    return {1: "I", 2: "II", 3: "III", 4: "IV", 5: "V"}.get(level or 0, str(level))

File: internal/usecase/eve/account.go
package eve

import (
	"context"
	"sort"
	"time"

	"github.com/truewebber/eve-online-mcp/internal/adapter/esi"
	"github.com/truewebber/eve-online-mcp/internal/domain/write"
	"github.com/truewebber/eve-online-mcp/internal/j"
	"github.com/truewebber/eve-online-mcp/internal/usecase/session"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

func registerAccount(s *mcp.Server) {
	addTool(s, &mcp.Tool{
		Name:        "eve_server_status",
		Description: "Tranquility server status: player count, build version, uptime, VIP mode.\n\nAlso the cheapest way to confirm this server can reach ESI at all. EVE has a daily downtime around 11:00 UTC; a low player count right after it is normal, not a bug.\n\nReturns: server_version, players, vip, start_time, data_age.",
	}, sessionTool(eveServerStatus))
	addTool(s, &mcp.Tool{
		Name:        "eve_auth_status",
		Description: "Who this connection is, and which in-game changes it can make.\n\nCall this before anything else when you do not know the setup, and always before promising the user an in-game change. This connection is signed in as exactly one character and acts only as that character; there is no way to add or switch to another one from here — that is a second server entry in the client, signed in separately.\n\nReturns: character, corporation, capabilities, capability_reference, outward_facing_capabilities, mails_last_hour, mails_remaining_this_hour, mail_cap_per_hour, pending_confirmations, confirm_ttl_seconds, confirm, session_expires_at.",
	}, sessionTool(eveAuthStatus))
	addTool(s, &mcp.Tool{
		Name:        "eve_auth_logout",
		Description: "Sign this connection out and revoke the server's access to its character.\n\nEnds the connection: the stored EVE authorization is revoked at CCP and every following call fails until the user signs in again through the browser. Destroys nothing in game. There is no argument — a connection can only log out the character it is.\n\nReturns: removed, character_id, character.",
	}, sessionTool(eveAuthLogout))
	addTool(s, &mcp.Tool{
		Name:        "eve_character_overview",
		Description: "Everything you would glance at on logging in: corp, ISK, location, ship, training.\n\nThe best first call for almost any question about how the character is doing — it fuses seven ESI endpoints into roughly 200 tokens and tells you what to drill into next. It already includes the wallet balance and what is training, so there is no need to ask for those separately.\n\nPartial results are normal: if one underlying endpoint fails, the rest still come back rather than the whole call erroring.\n\nReturns: name, corporation, alliance, security_status, wallet_isk, online, solar_system, docked_at, ship_type, training_now, queue_ends, remaps_available.",
	}, sessionTool(eveCharacterOverview))
}

func eveServerStatus(ctx context.Context, a *session.Session, _ empty) (any, error) {
	result, err := a.ESI.Get(ctx, esi.Path("/status"), nil, nil, nil)
	if err != nil {
		return nil, wrap("eveServerStatus", err)
	}
	out := j.Map(result.Data)
	out[fDataAge] = result.StaleNote()

	return out, nil
}

func eveAuthStatus(ctx context.Context, a *session.Session, _ empty) (any, error) {
	token, err := a.Character(ctx)
	if err != nil {
		return nil, wrap("eveAuthStatus", err)
	}
	policy := a.Guard.Status(ctx)
	var outward []string
	for name, cap := range write.Capabilities() {
		if cap.OutwardFacing {
			outward = append(outward, name)
		}
	}
	sort.Strings(outward)
	policy["outward_facing_capabilities"] = outward
	if live, err := a.Live(ctx); err == nil {
		policy["session_expires_at"] = live.ValidTil.UTC().Format(time.RFC3339)
	}
	writes := write.AllWriteScopeSet()
	corps := write.CorpScopeSet()

	return merge(map[string]any{
		"character": map[string]any{
			fName: token.CharacterName, fCharacterID: token.CharacterID,
			"scope_count":        len(token.Scopes),
			"write_scopes":       sortStrings(intersect(token.Scopes, writes)),
			"corporation_scopes": sortStrings(intersect(token.Scopes, corps)),
		},
	}, policy), nil
}

func eveAuthLogout(ctx context.Context, a *session.Session, _ empty) (any, error) {
	out, err := a.Logout(ctx)
	if err != nil {
		return nil, wrap("eveAuthLogout", err)
	}

	return map[string]any{"removed": out.CharacterName, fCharacterID: out.CharacterID}, nil
}

type overviewBox struct {
	r   esi.Result
	err error
}

func eveCharacterOverview(ctx context.Context, a *session.Session, _ empty) (any, error) {
	token, err := a.Character(ctx)
	if err != nil {
		return nil, wrap("eveCharacterOverview", err)
	}
	cid := token.CharacterID
	got := fetchOverview(ctx, a, cid)
	out := map[string]any{fCharacterID: cid, fName: token.CharacterName}
	applyOverviewPublic(ctx, a, got.public, out)
	applyOverviewWallet(got.wallet, out)
	applyOverviewOnline(got.online, out)
	applyOverviewLocation(ctx, a, cid, got.location, out)
	applyOverviewShip(ctx, a, got.ship, out)
	applyOverviewQueue(ctx, a, got.queue, out)
	if got.attributes.err == nil {
		out["remaps_available"] = j.Map(got.attributes.r.Data)["bonus_remaps"]
	}

	if age := staleOf(got.public, got.wallet, got.location, got.ship, got.online, got.queue, got.attributes); age != "" {
		out[fDataAge] = age
	}

	return compact(out), nil
}

func staleOf(boxes ...overviewBox) string {
	var ages []float64
	for _, b := range boxes {
		if b.err == nil {
			ages = append(ages, b.r.AgeSeconds)
		}
	}
	if len(ages) == 0 {
		return ""
	}

	return staleNote(ages...)
}

type overviewFetch struct {
	public, wallet, location, ship, online, queue, attributes overviewBox
}

func fetchOverview(ctx context.Context, a *session.Session, cid int) overviewFetch {
	get := func(path esi.Route, auth bool) overviewBox {
		var id *int
		if auth {
			id = &cid
		}
		r, err := a.ESI.Get(ctx, path, id, nil, nil)

		return overviewBox{r, err}
	}
	ch := make(chan func(*overviewFetch), overviewFetches)
	launch := func(path esi.Route, auth bool, set func(*overviewFetch, overviewBox)) {
		go func() {
			box := get(path, auth)
			ch <- func(out *overviewFetch) { set(out, box) }
		}()
	}
	launch(esiPath("characters", esiID(cid)), false, func(o *overviewFetch, b overviewBox) { o.public = b })
	launch(esiPath("characters", esiID(cid), fWallet), true, func(o *overviewFetch, b overviewBox) { o.wallet = b })
	launch(esiPath("characters", esiID(cid), fLocation), true, func(o *overviewFetch, b overviewBox) { o.location = b })
	launch(esiPath("characters", esiID(cid), fShip), true, func(o *overviewFetch, b overviewBox) { o.ship = b })
	launch(esiPath("characters", esiID(cid), "online"), true, func(o *overviewFetch, b overviewBox) { o.online = b })
	launch(esiPath("characters", esiID(cid), "skillqueue"), true, func(o *overviewFetch, b overviewBox) { o.queue = b })
	launch(esiPath("characters", esiID(cid), "attributes"), true, func(o *overviewFetch, b overviewBox) { o.attributes = b })

	var out overviewFetch
	for range overviewFetches {
		(<-ch)(&out)
	}

	return out
}

func applyOverviewPublic(ctx context.Context, a *session.Session, public overviewBox, out map[string]any) {
	if public.err != nil {
		return
	}
	info := j.Map(public.r.Data)
	ids := idsFrom(info["corporation_id"], info["alliance_id"])
	n, err := a.Resolver.Names(ctx, ids, nil)
	if err != nil {
		return
	}
	out[fCorporation] = n[j.Int(info["corporation_id"])]
	if j.Int(info["alliance_id"]) != 0 {
		out[fAlliance] = n[j.Int(info["alliance_id"])]
	}
	out["security_status"] = mathRound(j.Float(info["security_status"]), decimalPlaces)
	out["birthday"] = info["birthday"]
}

func applyOverviewWallet(wallet overviewBox, out map[string]any) {
	if wallet.err != nil || !isWalletAmount(wallet.r.Data) {
		return
	}
	out["wallet_isk"] = wallet.r.Data
	out[fWallet] = isk(wallet.r.Data)
}

func isWalletAmount(v any) bool {
	switch v.(type) {
	case float64, float32, int, int64:
		return true
	default:
		return false
	}
}

func applyOverviewOnline(online overviewBox, out map[string]any) {
	if online.err != nil {
		return
	}
	o := j.Map(online.r.Data)
	out["online"] = o["online"]
	out["last_login"] = o["last_login"]
}

func applyOverviewLocation(ctx context.Context, a *session.Session, cid int, location overviewBox, out map[string]any) {
	if location.err != nil {
		return
	}
	loc := j.Map(location.r.Data)
	placeIDs := idsFrom(loc["solar_system_id"], loc["station_id"], loc["structure_id"])
	n, err := a.Resolver.Names(ctx, placeIDs, &cid)
	if err != nil {
		return
	}
	out["solar_system"] = n[j.Int(loc["solar_system_id"])]
	docked := j.Int(loc["station_id"])
	if docked == 0 {
		docked = j.Int(loc["structure_id"])
	}
	if docked != 0 {
		out["docked_at"] = n[docked]
	} else {
		out["docked_at"] = "in space"
	}
	out["location_age"] = location.r.StaleNote()
}

func applyOverviewShip(ctx context.Context, a *session.Session, ship overviewBox, out map[string]any) {
	if ship.err != nil {
		return
	}
	sh := j.Map(ship.r.Data)
	name, err := a.Resolver.Name(ctx, j.Int(sh[fShipTypeID]), nil)
	if err != nil {
		return
	}
	out["ship_type"] = name
	if sn := j.Str(sh["ship_name"]); sn != "" && sn != name {
		out["ship_name"] = sn
	}
}

func applyOverviewQueue(ctx context.Context, a *session.Session, queue overviewBox, out map[string]any) {
	if queue.err != nil {
		return
	}
	if _, ok := queue.r.Data.([]any); !ok {
		return
	}
	var entries []map[string]any
	for _, e := range j.Maps(queue.r.Data) {
		if j.Str(e["finish_date"]) != "" {
			entries = append(entries, e)
		}
	}
	if len(entries) == 0 {
		out["warning"] = "Skill queue is empty — training time is being wasted."

		return
	}
	first := entries[0]
	skill, err := a.Resolver.Name(ctx, j.Int(first["skill_id"]), nil)
	if err != nil {
		return
	}
	out["training_now"] = skill + " " + roman(j.Int(first["finished_level"]))
	out["training_finishes"] = first["finish_date"]
	out["queue_length"] = len(entries)
	out["queue_ends"] = entries[len(entries)-1]["finish_date"]
}

func sortStrings(in []string) []string {
	out := append([]string{}, in...)
	sort.Strings(out)

	return out
}

func mathRound(v float64, places int) float64 {
	p := 1.0
	for range places {
		p *= 10
	}

	return float64(int(v*p+roundHalf)) / p
}
